csharp/tests/DeferTest.cs

```diff
@@ -0,0 +1,166 @@
+﻿// This file is part of the ArmoniK project
+//
+// Copyright (C) ANEO, 2022-2022. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License")
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using NUnit.Framework;
+
+namespace ArmoniK.Utils.Tests;
+
+public class DeferTest
+{
+  [Test]
+  public void DeferEmptyShouldWork()
+  {
+    using var defer = Deferrer.Empty;
+  }
+
+  [Test]
+  public void DeferShouldWork()
+  {
+    var i = 1;
+    using (Deferrer.Create(() => i += 1))
+    {
+      Assert.That(i,
+                  Is.EqualTo(1));
+    }
+
+    Assert.That(i,
+                Is.EqualTo(2));
+  }
+
+
+  [Test]
+  public void RedundantDeferShouldWork()
+  {
+    var i = 1;
+
+    var defer = Deferrer.Create(() => i += 1);
+
+    Assert.That(i,
+                Is.EqualTo(1));
+
+    defer.Dispose();
+
+    Assert.That(i,
+                Is.EqualTo(2));
+
+    defer.Dispose();
+
+    Assert.That(i,
+                Is.EqualTo(2));
+  }
+
+  [Test]
+  public async Task DeferShouldBeRaceConditionFree()
+  {
+    var i = 1;
+
+    var defer = Deferrer.Create(() =>
+                                {
+                                  Thread.Sleep(100);
+                                  i += 1;
+                                });
+
+    var task1 = Task.Run(() => defer.Dispose());
+    var task2 = Task.Run(() => defer.Dispose());
+
```

Comment:
Should check here that both tasks are still running :)

Reply:
It is both hard, and unnecessary. The value on `i` at the end is enough to check that there is no race condition.